evm/src/lookup.rs

```diff
@@ -1,127 +1,248 @@
-    vars: StarkEvaluationTargets<D, COLS>,
-    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
-    col_permuted_input: usize,
-    col_permuted_table: usize,
-) {
-    let local_perm_input = vars.local_values[col_permuted_input];
-    let next_perm_table = vars.next_values[col_permuted_table];
-    let next_perm_input = vars.next_values[col_permuted_input];
-
-    // A "vertical" diff between the local and next permuted inputs.
-    let diff_input_prev = builder.sub_extension(next_perm_input, local_perm_input);
-    // A "horizontal" diff between the next permuted input and permuted table value.
-    let diff_input_table = builder.sub_extension(next_perm_input, next_perm_table);
-
-    let diff_product = builder.mul_extension(diff_input_prev, diff_input_table);
-    yield_constr.constraint(builder, diff_product);
-
-    // This is actually constraining the first row, as per the spec, since `diff_input_table`
-    // is a diff of the next row's values. In the context of `constraint_last_row`, the next
-    // row is the first row.
-    yield_constr.constraint_last_row(builder, diff_input_table);
+impl Lookup {
+    pub(crate) fn num_helper_columns(&self, constraint_degree: usize) -> usize {
+        // One helper column for each column batch of size `constraint_degree-1`,
+        // then one column for the inverse of `table + challenge` and one for the `Z` polynomial.
+        ceil_div_usize(self.columns.len(), constraint_degree - 1) + 2
+    }
 }
 
-/// Given an input column and a table column, generate the permuted input and permuted table columns
-/// used in the Halo2 permutation argument.
-pub fn permuted_cols<F: PrimeField64>(inputs: &[F], table: &[F]) -> (Vec<F>, Vec<F>) {
-    let n = inputs.len();
+/// logUp protocol from https://ia.cr/2022/1530 (TODO link to newer version?)
+/// Compute the helper columns for the lookup argument.
+/// Given columns `f0,...,fk` and a column `t`, such that `∪fi ⊆ t`, and challenges `x`,
+/// this computes the helper columns `h_i = 1/(x+f_2i) + 1/(x+f_2i+1)`, `g = 1/(x+t)`,
+/// and `Z(gx) = Z(x) + sum h_i(x) - m(x)g(x)` where `m` is the frequencies column.
+pub(crate) fn lookup_helper_columns<F: Field>(
+    lookup: &Lookup,
+    trace_poly_values: &[PolynomialValues<F>],
+    challenge: F,
+    constraint_degree: usize,
+) -> Vec<PolynomialValues<F>> {
+    assert_eq!(
+        constraint_degree, 3,
+        "TODO: Allow other constraint degrees."
+    );
+    let num_helper_columns = lookup.num_helper_columns(constraint_degree);
+    let mut helper_columns: Vec<PolynomialValues<F>> = Vec::with_capacity(num_helper_columns);
+
+    // For each batch of `constraint_degree-1` columns `fi`, compute `sum 1/(f_i+challenge)` and
+    // add it to the helper columns.
+    // TODO: This does one batch inversion per column. It would also be possible to do one batch inversion
+    // for every column, but that would require building a big vector of all the columns concatenated.
```

Comment:
I suppose that you want to say "It would be possible to do one batch inversion for every *group/block* of columns.... "